kotori/daq/storage/cratedb.py

```diff
@@ -0,0 +1,270 @@
+
+        except Exception as ex:
+            log.failure(
+                "Could not format chunk (ex={ex_name}: {ex}): data={data}, meta={meta}",
+                ex_name=ex.__class__.__name__,
+                ex=ex,
+                meta=meta_copy,
+                data=data_copy,
+            )
+            raise
+
+        try:
+            success = self.write_chunk(meta, chunk)
+            return success
+
+        except requests.exceptions.ConnectionError as ex:
+            log.failure(
+                "Problem connecting to CrateDB at {uri}: {ex}", uri=self.host_uri, ex=ex
+            )
+            raise
+
+        except ProgrammingError as ex:
+            if "SchemaUnknownException" in ex.message:
+                db_table = self.get_tablename(meta)
+                self.create_table(db_table)
+
+                # Attempt second write
+                success = self.write_chunk(meta, chunk)
+                return success
+
+            else:
+                raise
+
+    def write_chunk(self, meta, chunk):
+        """
+        Run the SQL `INSERT` operation.
+        """
+        db_table = self.get_tablename(meta)
+        cursor = self.db_client.cursor()
+
+        # With or without timestamp.
+        if "time" in chunk:
+            cursor.execute(
+                f"INSERT INTO {db_table} (time, tags, fields) VALUES (?, ?, ?)",
+                (chunk["time"], chunk["tags"], chunk["fields"]),
+            )
+        else:
+            cursor.execute(
+                f"INSERT INTO {db_table} (tags, fields) VALUES (?, ?)",
+                (chunk["tags"], chunk["fields"]),
+            )
+        success = True
+        self.databases_written_once.add(meta.database)
+        cursor.close()
+        if success:
+            log.debug("Storage success: {chunk}", chunk=chunk)
+        else:
+            log.error("Storage failed:  {chunk}", chunk=chunk)
+        return success
+
+    @staticmethod
+    def get_tags(data):
+        """
+        Derive tags from topology information.
+
+        TODO: Verify if this is used at all.
+        """
+        return project(data, ["gateway", "node"])
+
+
+class TimezoneAwareCrateJsonEncoder(json.JSONEncoder):
+    epoch_aware = datetime(1970, 1, 1, tzinfo=pytz.UTC)
+    epoch_naive = datetime(1970, 1, 1)
+
+    def default(self, o):
+        if isinstance(o, Decimal):
+            return str(o)
+        if isinstance(o, datetime):
+            if o.tzinfo:
+                delta = o - self.epoch_aware
+            else:
+                delta = o - self.epoch_naive
+            return int(delta.microseconds / 1000.0 +
+                       (delta.seconds + delta.days * 24 * 3600) * 1000.0)
+        if isinstance(o, date):
+            return calendar.timegm(o.timetuple()) * 1000
+        return json.JSONEncoder.default(self, o)
+
+
+# Monkey patch.
+# TODO: Submit upstream.
+crate.client.http.CrateJsonEncoder = TimezoneAwareCrateJsonEncoder
```

Comment:
This patch should be submitted upstream to the Python driver on behalf of the `crate` Python package. It may resolve https://github.com/crate/crate-python/issues/361.